### PR TITLE
kubeadm: backup kubelet config for "upgrade node" and "upgrade apply"

### PR DESCRIPTION
**Pull request by chendave**
- cherry pick the previous commit which is reverted by [revert feature to backup kubelet config for "upgrade node"](https://github.com/kubernetes/kubernetes/pull/114608), co-authored by @pacoxu 
- fix `invalid cross-device link` error, for the issue first reported here https://github.com/kubernetes/kubeadm/issues/2796 and the journal from here: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-patches-latest/1605265548596219904/build-log.txt
The root cause for that error is because `rename` doesn't work across different mount points.
The kubelet config file and back up directory are mounted to different file system in kinder environment.
    ```
    df /var/lib/kubelet/config.yaml | tail -n1 | awk '{print $1}'
    /dev/sda2
    df /etc/kubernetes/tmp/kubeadm-kubelet-configxxx | tail -n1 | awk '{print $1}'
    overlay
    ```
    Call `cp` instead of `rename` to back up the kubelet file would fix that issue.

- include the err got from `PrintDryRunFile`
The error was ingored which means if anything wrong from `PrintDryRunFiles`, it was sliently ignored.

- backup kubelet config file for `upgrade apply`
Backup kubelet config file for `kubeadm upgrade apply`, some code refactoring is done to de-dup some redundant code logic.

test result with kinder: https://gist.github.com/chendave/64f7825cf7faa7f30774fe6c48053c21

/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
kubelet config file will be backed up to `/etc/kubernetes/tmp/` folder with `kubeadm-kubelet-config` append with a random suffix as the filename 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by chendave** [comment]
/hold
need more self-testing, but I believe this pr will fix the issue we found.

/cc @neolit123  @pacoxu

**Comment by chendave** [comment]
- here is the test result with kinder: https://gist.github.com/chendave/64f7825cf7faa7f30774fe6c48053c21
- update the pr with testcases created.

/hold cancel
this is ready for review now.

**Comment by pacoxu** [comment]
I am working on some old PR follow-ups this week and will update some KEPs in this week and next. I will get back to this later this week or next then. 
/assign

**Comment by chendave** [comment]
need a rebase since https://github.com/kubernetes/kubernetes/pull/114719 was merged.

**Comment by chendave** [comment]
@pacoxu @neolit123 @SataQiu might out of office and only a few availability in the following weeks, pls let me know if there is anything more I can do for this, thanks!

File: cmd/kubeadm/app/phases/upgrade/postupgrade_test.go [hunk 1]
```diff
@@ -101,3 +104,98 @@ func TestRollbackFiles(t *testing.T) {
 		t.Fatalf("Expected error contains %q, got %v", errString, err)
 	}
 }
+
+func TestWriteKubeletConfigFiles(t *testing.T) {
```
Comment (pacoxu) [inline cmd/kubeadm/app/phases/upgrade/postupgrade_test.go:108]:
```
--- FAIL: TestWriteKubeletConfigFiles (0.00s)
    /Users/pacoxu/git/gopath/src/k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade/postupgrade_test.go:155: Expected error contains "no kubelet component config found.*no such file or directory", got failed to create directory "/etc/kubernetes/tmp": mkdir /etc/kubernetes: permission denied
FAIL
FAIL	k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade	1.058s
FAIL
```
This UT will fail on Mac for permission.

Comment (chendave) [reply cmd/kubeadm/app/phases/upgrade/postupgrade_test.go:108]: If this is a problem on the Mac, it should be a long-standing issues, as the `CreateTempDirForKubeadm` is called in many places. 

Have you run the test with `root`? 

Let me find a Mac and try it out, anyway, if it is an issue, the issue should be fixed separately.

File: cmd/kubeadm/app/cmd/phases/upgrade/node/kubeletconfig.go [hunk 2]
```diff
@@ -65,41 +60,13 @@ func runKubeletConfigPhase() func(c workflow.RunData) error {
...
-		if dryRun {
-			if err := printFilesIfDryRunning(dryRun, kubeletDir); err != nil {
-				return errors.Wrap(err, "error printing files on dryrun")
-			}
-			return nil
-		}
+		// Write the configuration for the kubelet down to disk and print the generated manifests instead if dry-running.
+		// If not dry-running, the kubelet config file will be backed up to /etc/kubernetes/tmp/ dir, so that it could be
+		// recovered if there is anything goes wrong.
+		upgrade.WriteKubeletConfigFiles(cfg, data.PatchesDir(), dryRun, data.OutputWriter())
```
Comment (pacoxu) [inline cmd/kubeadm/app/cmd/phases/upgrade/node/kubeletconfig.go:66]: Should we handle the return err  here?
```
err := upgrade.WriteKubeletConfigFiles(cfg, data.PatchesDir(), dryRun, data.OutputWriter())
```

**Comment by chendave** [comment]
@pacoxu pls see https://github.com/kubernetes/kubernetes/compare/f6b04fe5d4e168067d978440af4464134f57caa4..484d0d542e3adbcbda08a48236daabcfe1649504 for the diff.

added the check for privileged user so that the test can only run with the root user.  Your comments leads me to think about the regression we have: https://github.com/kubernetes/kubernetes/issues/114824 👍

File: cmd/kubeadm/app/cmd/phases/upgrade/node/kubeletconfig.go (hunk 2 above)
Comment (chendave) [reply cmd/kubeadm/app/cmd/phases/upgrade/node/kubeletconfig.go]: yes! addressed.

**Comment by pacoxu** [comment]
/priority important-soon
/triage accepted

> https://github.com/kubernetes/kubernetes/compare/f6b04fe5d4e168067d978440af4464134f57caa4..484d0d542e3adbcbda08a48236daabcfe1649504 for the diff.
> 
> added the check for privileged user so that the test can only run with the root user. Your comments leads me to think about the regression we have: #114824 👍

I am not sure if this needs to be fixed or not, and is this the best practice to skip it like this?

Overall
/lgtm

**Comment by chendave** [comment]
> I am not sure if this needs to be fixed or not, and is this the best practice to skip it like this?

my understanding is that upgrade needs the user to be root, so it is non-sense to run those test if the user is not root, besides, I was thinking to create the dir which *every* user has the permission to write on all the arches & os, but I don't think its possible.

**Comment by neolit123** [comment]
/hold
/approve

please test with kinder and the related e2e test workflow before merging / removing hold.
docs on how to use kinder can be found in the k/kubeadm repo

**Comment by neolit123** [comment]
actually i realize it's not so simple to test uncommited code with kinder....that is because the tool is mostly designed for testing artifacts that are already build from k/k commits in CI.

alternatively just remove the hold watch the e2e test signal  in the next 2 days and revert / update if needed:
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-upgrade-1-22-1-23

**Comment by chendave** [comment]
@neolit123  I tested it locally, and here is the link of the result I pasted in the desc,

> test result with kinder: https://gist.github.com/chendave/64f7825cf7faa7f30774fe6c48053c21

**Comment by chendave** [comment]
/hold cancel

let's see if the CI job is happy.

**Comment by pacoxu** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-patches-latest/1622937946548080640
This passed after merge.

**Pull request merged**